confluentinc/ksql
# feat: move command topic deserialization to CommandRunner and introduce DEGRADED CommandRunnerStatus

### Description 
Moves deserialization of the ConsumerRecords from the CommandTopic to the CommandRunner. This allows for deserialization of individual records and better detection of `SerializationException` that can occur when a server processes an incompatible command.

Introduces a `DEGRADED` `CommandRunnerStatus` which indicates that the CommandRunner has encountered an incompatible process and the thread has been terminated. The only way to get out of this state currently is to start up a server version that can process the command that put old server version into the `DEGRADED` state

Next step is to add a check in DistributingExecutor that makes sure the CommandRunner isn't in a DEGRADED state.

### Testing done 
Updated Unit tests and added new ones
Manual test:

- Start with fresh command topic
- Start up server on this branch
- Issue a bunch of statements to the command topic (CS/CSAS)
- Stopped the server
- Created a new branch and added a field to the Command object
- Start server and issue DDL statements
- Switched to this branch and start up server
- Verify with JConsole the CommandRunnerStatus is degraded
- Verify all previous commands before the incompatible command are executed and the queries are started
- Verify CommandRunner thread is shut down by trying to issue more DDL statements to server

### Reviewer checklist
